### PR TITLE
New User defined components class

### PR DESCRIPTION
**Pull request by Samuelopez-ansys**
- Class to obtain 3DComponent and User Defined Model information
- Getter and setter object
- "copy_solids_from" method accepting UserDefinedComponent objects

### CODE REVIEW

File: pyaedt/modeler/Object3d.py [hunk 1]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+
+        """
+        group = None
+        if "Group" in self._primitives.oeditor.GetChildObject(self.name).GetPropNames():
+            group = self._primitives.oeditor.GetChildObject(self.name).GetPropValue("Group")
+        return group
+
+    @group_name.setter
+    def group_name(self, name):
+        """Assign Component to a specific group. it creates a new group if the group doesn't exist.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:3858]:
```suggestion
        """Assign component to a specific group. A new group is created if the specified group doesn't exist.
```

File: pyaedt/modeler/Object3d.py [hunk 2]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+
+        Parameters
+        ----------
+        name : str
+            Name of the group to assign. Group will be created if it does not exist.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:3863]:
```suggestion
            Name of the group to assign the component to. A group is created if the one
            specified does not exist.
```

File: pyaedt/modeler/Object3d.py [hunk 3]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+        return False
+
+    @property
+    def mesh_assembly(self):
+        """Mesh assembly flag.
+
+        Returns
+        -------
+        bool
+           ``True`` if Mesh assembly checked, ``None`` if user defined model
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:3932]:
```suggestion
           ``True`` if mesh assembly is checked, ``None`` if a user-defined model.
```

File: pyaedt/modeler/Object3d.py [hunk 4]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+            self.is3dcomponent
+            and isinstance(ma, bool)
+            and key in self._primitives.oeditor.GetChildObject(self.name).GetPropNames()
+        ):
+            self._primitives.oeditor.GetChildObject(self.name).SetPropValue(key, ma)
+            self._mesh_assembly = ma
+
+    @property
+    def name(self):
+        """Name of the object as a string value.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:3956]:
```suggestion
        """Name of the object.
```

File: pyaedt/modeler/Object3d.py [hunk 5]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+
+        Returns
+        -------
+        str
+           Name of object as a string value.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:3961]:
```suggestion
           Name of object.
```

File: pyaedt/modeler/Object3d.py [hunk 6]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+            )
+            if tCS in tCS_options:
+                pcs = ["NAME:Target Coordinate System", "Value:=", tCS]
+                self._change_property(pcs)
+                self._target_coordinate_system = tCS
+
+    @pyaedt_function_handler()
+    def delete(self):
+        """Delete the object. Project must be saved after the operation in order to update the user defined component
+        names list.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:4078]:
```suggestion
        """Delete the object. Project must be saved after the operation to update the list
        of names for user-defined components.
```

File: pyaedt/modeler/Object3d.py [hunk 7]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+
+    @pyaedt_function_handler()
+    def duplicate_around_axis(self, cs_axis, angle=90, nclones=2, create_new_objects=True):
+        """Duplicate the component around the axis.
+
+        Parameters
+        ----------
+        cs_axis : Application.CoordinateSystemAxis object
+            Coordinate system axis of the object.
+        angle : float
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:4210]:
```suggestion
        angle : float, optional
```

File: pyaedt/modeler/Object3d.py [hunk 8]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+
+        >>> oEditor.DuplicateAlongLine
+
+        """
+        if self.is3dcomponent:
+            _, added_objects = self._primitives.modeler.duplicate_along_line(
+                self.name, vector, nclones, attachObject, True
+            )
+            return added_objects
+        self._logger.warning("User defined models does not support this operation")
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:4265]:
```suggestion
        self._logger.warning("User-defined models do not support this operation.")
```

File: pyaedt/modeler/Primitives.py [hunk 9]
```diff
@@ -996,7 +998,7 @@ def object_names(self):
         return self._all_object_names
 
     @property
-    def components_3d_names(self):
+    def user_defined_component_names(self):
         """List of the names of all 3d components objects.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives.py:1002]:
```suggestion
        """List of the names of all 3D components objects.
```

File: pyaedt/modeler/Primitives.py [hunk 10]
```diff
@@ -1981,13 +2006,32 @@ def add_new_objects(self):
                 added_objects.append(obj_name)
         return added_objects
 
+    @pyaedt_function_handler()
+    def add_new_user_defined_component(self):
+        """Add 3DComponents and User defined models that have been created in the modeler by
```
Comment (PipKat) [inline pyaedt/modeler/Primitives.py:2011]:
```suggestion
        """Add 3D components and user-defined models that have been created in the modeler by
```

File: pyaedt/modeler/Primitives.py [hunk 11]
```diff
@@ -3523,6 +3573,30 @@ def _find_object_from_face_id(self, lval):
 
         return None
 
+    @pyaedt_function_handler()
+    def _get_user_defined_component_names(self):
+        """Retrieve user defined models list names.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives.py:3578]:
```suggestion
        """Get list of names for user-defined models.
```

File: pyaedt/modeler/Primitives.py [hunk 12]
```diff
@@ -3523,6 +3573,30 @@ def _find_object_from_face_id(self, lval):
...
+
+        Returns
+        -------
+        [List with User Defined Model names]
```
Comment (PipKat) [inline pyaedt/modeler/Primitives.py:3582]: list
    List of names for user-defined models.

Comment (PipKat) [inline pyaedt/modeler/Primitives.py:3582]:
```suggestion
       list
           List of names for user-defined models.
```

File: pyaedt/modeler/Primitives.py [hunk 13]
```diff
@@ -3523,6 +3573,30 @@ def _find_object_from_face_id(self, lval):
...
+                    "UserDefinedModels"
+                ]
+                if entity_list:
+                    udm_entry = entity_list["UserDefinedModel"]
+                    if isinstance(udm_entry, (dict, OrderedDict)):
+                        udm_entry = [udm_entry]
+                    for data in udm_entry:
+                        udm_lists.append(data["Attributes"]["Name"])
+            except:
+                self.logger.error("User Defined Models were not retrieved from AEDT file")
```
Comment (PipKat) [inline pyaedt/modeler/Primitives.py:3597]:
```suggestion
                self.logger.error("User-defined models were not retrieved from the AEDT file.")
```

File: pyaedt/modeler/Primitives.py [hunk 14]
```diff
@@ -3523,6 +3573,30 @@ def _find_object_from_face_id(self, lval):
...
+        return udm_lists
+
     @pyaedt_function_handler()
     def __getitem__(self, partId):
         """Return the object ``Object3D`` for a given object ID or object name.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives.py:3602]:
```suggestion
        """Get the object ``Object3D`` for a given object ID or object name.
```

File: pyaedt/modeler/Primitives3D.py [hunk 15]
```diff
@@ -993,11 +999,13 @@ def create_udm(self, udmfullname, udm_params_list, udm_library="syslib"):
             List of user-defined object pairs for the model.
         udm_library : str, optional
             Name of library for the user-defined model. The default is ``"syslib"``.
+        name : str, optional
+            User defined model name. Default is `None`.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3D.py:1003]:
```suggestion
            Name of the user-defined model. The default is ``None```.
```

File: pyaedt/modeler/Primitives3D.py [hunk 16]
```diff
@@ -993,11 +999,13 @@ def create_udm(self, udmfullname, udm_params_list, udm_library="syslib"):
             List of user-defined object pairs for the model.
         udm_library : str, optional
             Name of library for the user-defined model. The default is ``"syslib"``.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3D.py:1001]:
```suggestion
            Name of the library for the user-defined model. The default is ``"syslib"``.
```

File: pyaedt/modeler/Primitives3D.py [hunk 17]
```diff
@@ -993,11 +999,13 @@ def create_udm(self, udmfullname, udm_params_list, udm_library="syslib"):
...
+        name : str, optional
+            User defined model name. Default is `None`.
 
         Returns
         -------
-        :class:`pyaedt.modeler.Object3d.Object3d`
-            3D object.
+        :class:`pyaedt.modeler.Object3d.UserDefinedComponent`
+            User defined component object.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3D.py:1008]:
```suggestion
            User-defined component object.
```

File: pyaedt/modeler/Primitives3D.py [hunk 18]
```diff
@@ -1144,11 +1163,13 @@ def insert_3d_component(self, compFile, geoParams=None, szMatParams="", szDesign
             Design parameters. The default is ``""``.
         targetCS : str, optional
             Target coordinate system. The default is ``"Global"``.
+        name : str, optional
+            3D component name. Default is `None`.
```
Comment (PipKat) [inline pyaedt/modeler/Primitives3D.py:1167]:
```suggestion
            3D component name. The default is ``None``.
```

File: pyaedt/modeler/Object3d.py [hunk 19]
```diff
@@ -3750,3 +3750,560 @@ def change_property(self, vPropChange, names_list=None):
...
+            the Application.Position object for the selection.
+        vector : float
+            List of the ``[x1, y1, z1]`` coordinates or
+            the Application.Position object for the vector.
+
+        Returns
+        -------
+        pyaedt.modeler.Object3d.UserDefinedComponent, bool
+            3D object.
+            ``False`` when failed.
```
[the remark this answers is not in the document]
Comment (maxcapodi78) [reply pyaedt/modeler/Object3d.py:4119]: it returns the object if True, False when it fails

File: pyaedt/modeler/Object3d.py [hunk 20]
```diff
@@ -3750,3 +3750,561 @@ def change_property(self, vPropChange, names_list=None):
...
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:4120]:
```suggestion
            3D object when successful, ``False`` when failed.
```

File: pyaedt/modeler/Object3d.py [hunk 21]
```diff
@@ -3750,3 +3750,561 @@ def change_property(self, vPropChange, names_list=None):
...
+            degrees or radians. The default is ``90.0``.
+        unit : text, optional
+             Units for the angle. Options are ``"deg"`` or ``"rad"``.
+             The default is ``"deg"``.
+
+        Returns
+        -------
+        pyaedt.modeler.Object3d.UserDefinedComponent, bool
+            3D object.
+            ``False`` when failed.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:4155]:
```suggestion
            3D object when successful, ``False`` when failed.
```

File: pyaedt/modeler/Object3d.py [hunk 22]
```diff
@@ -3750,3 +3750,561 @@ def change_property(self, vPropChange, names_list=None):
...
+            List of object IDs.
+        vector : list
+            Vector of the direction move. It can be a list of the ``[x, y, z]``
+            coordinates or a Position object.
+
+        Returns
+        -------
+        pyaedt.modeler.Object3d.UserDefinedComponent, bool
+            3D object.
+            ``False`` when failed.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:4187]:
```suggestion
            3D object when successful, ``False`` when failed.
```

**Pull request merged**